src/new_api_v04/constraint.rs

```diff
@@ -76,43 +97,74 @@ where
     AP: AssignPredicate<U = U>,
     FP: FilterPredicate<U = U>,
 {
-    /// Construct a constraint.
+    /// Construct a binary constraint.
     ///
     /// Returns an error if the constraint is malformed, i.e. if the predicate is
-    /// a Predicate::Assign and the object is not a variable.
-    pub fn try_from_triple(
+    /// a Predicate::Assign and the object is not a variable, or if the predicate
+    /// is not binary.
+    pub fn try_binary_from_triple(
         lhs: ConstraintLiteral<V, U>,
         predicate: Predicate<AP, FP>,
         rhs: ConstraintLiteral<V, U>,
     ) -> Result<Self, InvalidConstraint> {
+        if predicate.arity() != 2 {
+            return Err(InvalidConstraint::InvalidArity);
+        }
         if matches!(predicate, Predicate::Assign(_)) && matches!(rhs, ConstraintLiteral::Value(_)) {
             return Err(InvalidConstraint::AssignToValue);
         }
-        Ok(Self {
-            lhs,
-            predicate,
-            rhs,
-        })
+        Self::try_new(predicate, vec![lhs, rhs])
+    }
+
+    /// Construct a constraint.
+    ///
+    /// Returns an error if the number of arguments does not match the predicate
+    /// arity.
+    pub fn try_new(
+        predicate: Predicate<AP, FP>,
+        args: Vec<ConstraintLiteral<V, U>>,
+    ) -> Result<Self, InvalidConstraint> {
+        if args.len() != predicate.arity() {
+            Err(InvalidConstraint::InvalidArity)
+        } else if matches!(predicate, Predicate::Assign(_))
+            && matches!(args.last(), Some(ConstraintLiteral::Value(_)))
+        {
+            Err(InvalidConstraint::AssignToValue)
+        } else {
+            Ok(Self { args, predicate })
+        }
+    }
+
+    /// Return the arity of the constraint.
+    pub fn arity(&self) -> usize {
+        let arity = self.predicate.arity();
+        assert_eq!(self.args.len(), arity, "invalid constraint: arity mismatch");
+        arity
```

Comment:
This is rather a `debug_assert_eq`, as it should be checked at construction time.
Also, `Vec::len` is the faster option.

```suggestion
    /// Return the arity of the constraint.
    pub fn arity(&self) -> usize {
        let arity = self.args.len();
        debug_assert_eq!(arity, self.predicate.arity(), "invalid constraint: arity mismatch");
        arity
```